### PR TITLE
DAOS-4590 mgmt: change daos_mgmt_svc_rip() to fork dmg

### PR DESCRIPTION
**Pull request by GitHuaKuang**
As one step to remove src/include/daos_mgmt.h and its APIs, changed daos_mgmt_svc_rip()
to system call the pre-built dmg command.

Change-Id: I7fb9f6dcc01dbe9eab6f27b5f1fb3cc2d9b42fe7
Signed-off-by: Hua Kuang <hua.kuang@intel.com>

### CODE REVIEW

**Comment by mjmac** [comment]
@GitHuaKuang: It would be great if you could merge latest master to pull in test fixes so that you're more likely to get a clean test run.

**Comment by GitHuaKuang** [comment]
@johannlombardi My understanding of the goal is to replace all APIs in daos_mgmt.h to use dmg cmd, then we can remove this header file.

Currently most of these APIs are called by daos_test in multiple places, instead of doing the fork (you can see the code needed in this PR) each call in tests/suite, @mjmac  suggested and I agreed that keeping a wrapper so tests/suite code can be minimized, in this 1st PR, I even kept the same API as the wrapper, but its inside is doing dmg forking.

We can change the wrapper names(existing APIs)  to reflect dmg_fork_xxxx() if that fits the big plan better.

**Comment by mjmac** [comment]
Hmm. I assumed that this was the beginning of a prototype rather than something to merge into master. When I suggested a compatibility shim, I meant a new file used only by daos_test, not a modification to the client API. I think the goal is to remove those symbols completely from the public API.

**Comment by GitHuaKuang** [comment]
OK, let me remodel it to a daos_test only version shim to do the dmg fork.

File: src/client/api/mgmt.c [hunk 1]
```diff
@@ -22,32 +22,78 @@
...
+	pid_t pid = fork();
+
+	if (pid == -1) {
+		D_ERROR("fork() failed to create child\n");
+		return -DER_INVAL;
+	} else if (pid == 0) {
+		/* child process */
+		D_INFO("child pid = %d\n", getpid());
+		execvp(argv[0], argv);
+		exit(DER_SUCCESS);
```
Comment (jolivier23) [inline src/client/api/mgmt.c:45]: execvp should not return.  If we get here, something went wrong and we should return an error.

File: src/common/tests_lib.c [hunk 2]
```diff
@@ -355,3 +356,71 @@ dts_sgl_init_with_strings_repeat(d_sg_list_t *sgl, uint32_t repeat,
...
+		if (waitpid(pid, &status, 0) == -1) {
+			D_ERROR("waitpid() failed for pid = %d\n", pid);
+			return -DER_INVAL;
+		}
+		if (WIFEXITED(status))
+			D_INFO("WEXITSTATUS(status) = %d\n",
+				WEXITSTATUS(status));
+		return DER_SUCCESS;
+	}
+}
```
Comment (johannlombardi) [inline src/common/tests_lib.c:390]: Why not just use the system(3) function provided by the libc?

Comment (GitHuaKuang) [reply src/common/tests_lib.c]: right, system call does most what I am trying to do via fork/exec here, using it, now the code change becomes very small, so we don't even need a shim anymore, see the new patch.

File: src/tests/suite/daos_test_common.c [hunk 3]
```diff
@@ -898,7 +899,10 @@ daos_kill_server(test_arg_t *arg, const uuid_t pool_uuid,
 		      "disabled, svc->rl_nr %d)!\n", rank, arg->srv_ntgts,
 		       arg->srv_disabled_ntgts - 1, svc->rl_nr);
 
-	rc = daos_mgmt_svc_rip(grp, rank, true, NULL);
+	/* build and invoke dmg cmd to stop the server */
+	snprintf(dmg_cmd, sizeof(dmg_cmd),
```
Comment (jolivier23) [inline src/tests/suite/daos_test_common.c:903]: You need to check the return code.   Alternatively, you could use dts_create_config here instead and make buffer use DTS_CFG_MAX as the size.   This function checks the return code and ensures the string is null terminated.

Comment (jolivier23) [reply src/tests/suite/daos_test_common.c:903]: Admittedly, this function is poorly named for what you're doing...but it does the same thing

**Comment by GitHuaKuang** [comment]
It just passed all CI tests for the 1st time with dmg cmd, please take look see if ready for landing.

File: src/tests/suite/daos_test_common.c [hunk 4]
```diff
@@ -899,8 +902,19 @@ daos_kill_server(test_arg_t *arg, const uuid_t pool_uuid,
 		      "disabled, svc->rl_nr %d)!\n", rank, arg->srv_ntgts,
 		       arg->srv_disabled_ntgts - 1, svc->rl_nr);
 
-	rc = daos_mgmt_svc_rip(grp, rank, true, NULL);
-	assert_int_equal(rc, 0);
+	fp = popen("which dmg", "r");
```
Comment (jolivier23) [inline src/tests/suite/daos_test_common.c:905]: what is the purpose of this?   Either dmg is in the path or it isn't.   Why do we need to resolve the full path before creating the command?  I would just remove this block and use dmg.   Regardless, I don't think calling which is the right strategy here.

Comment (GitHuaKuang) [reply src/tests/suite/daos_test_common.c:905]: It is a not clean handling - several tests failed with  system(dmg) return code 0x100, likely is not able to locate dmg. After passing the full path, the test worked.

Not clear how dmg path is define in CI test in this daos_test C code, I know Avocado code set it in python using .build_vars.json, Wondering we need to do something similar in C code.

Comment (GitHuaKuang) [reply src/tests/suite/daos_test_common.c:905]: This was the test wihtou "which dmg" rebase 4 days ago that had more failures with 0x100 return. Let me re-test with out full path, hopefully the failures were caused by other issues which is addressed.
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2620/13/tests

Comment (GitHuaKuang) [reply src/tests/suite/daos_test_common.c]: It passed CI functional test without full path.

**Comment by GitHuaKuang** [comment]
@sydneyvanda @wangdi1  Can you give a review to this? Thanks.

File: src/tests/suite/daos_test_common.c [hunk 5]
```diff
@@ -899,8 +900,14 @@ daos_kill_server(test_arg_t *arg, const uuid_t pool_uuid,
...
 		       arg->srv_disabled_ntgts - 1, svc->rl_nr);
 
-	rc = daos_mgmt_svc_rip(grp, rank, true, NULL);
-	assert_int_equal(rc, 0);
+	/* build and invoke dmg cmd to stop the server */
+	dts_create_config(dmg_cmd, "dmg system stop -i --ranks=%d --force",
+			  rank);
+	rc = system(dmg_cmd);
+	if (rc == -1 || WEXITSTATUS(rc) != 0) {
+		print_message("%s failed rc=%d\n", dmg_cmd, rc);
```
Comment (jolivier23) [inline src/tests/suite/daos_test_common.c:908]: should this not be an assert (or at least return an error to the caller)?   The prio code asserted

**Comment by GitHuaKuang** [comment]
@daosbuild1 Can we take another review on this? Since it added dmg config file to daos_test which is needed for all other tickets in this group, need it to be landed before others API changes can be pushed.

**Comment by mjmac** [comment]
It seems fine to me, but I'm not really a stakeholder in this area... I think someone from the test team (@phender, etc) should take a look at it.

**Pull request merged**